node/overseer/src/lib.rs

```diff
@@ -884,14 +889,20 @@ where
 	fn handle_external_request(&mut self, request: ExternalRequest) {
 		match request {
 			ExternalRequest::WaitForActivation { hash, response_channel } => {
-				// We use known leaves here because the `WaitForActivation` message
-				// is primarily concerned about leaves which subsystems have simply
-				// not been made aware of yet. Anything in the known leaves set,
-				// even if stale, has been activated in the past.
-				if self.known_leaves.peek(&hash).is_some() {
```

Comment:
This comment still seems accurate. It seems that the change would require the leaf to become activated again. Does that actually happen?

Reply:
Yes, on reverts: or I am totally misunderstanding the whole Stale/Fresh leaf mechanism. Is [BlockImported](https://github.com/paritytech/polkadot/blob/fd31407737999e8f1663799055e76b0efd3611cc/node/overseer/src/lib.rs#L735) triggered in that case? It should, otherwise [this](https://github.com/paritytech/polkadot/blob/fd31407737999e8f1663799055e76b0efd3611cc/node/overseer/src/lib.rs#L861) would be useless.

I was about to say that we already tested this, but did we? Chain reversions did work, but I guess they would also work if we did not get a `BlockImported` event. (We would just not provide any ParasInherent) .... So not so sure any more. 

Anyhow, even if we did not receive a `BlockImported` event, then we would not have a leaf at all and could not handle the create inherent call anyway. (This is where we use the `ExternalRequest::WaitForActivation` at the moment.)


About the comment: The activation in the past is of no value, as it might (and likely has been) deactivated already again. We only know it is active _now_ if it is in `active_leaves`.

Reply:
Given https://github.com/paritytech/substrate/issues/12375#issuecomment-1260736152, how does that even work if we remove leaves on deactivation in provisioner or did I misunderstand something?

To me it seems like we want to revert this change, but be more lenient with deactivation in provisioner.

Reply:
That issue is unrelated to this PR. Not deactivating (the relevant deactivation was in backing) would only let us do more pointless work.

Reply:
I meant this comment
> We [creating the inherents for the parent block](https://github.com/paritytech/substrate/blob/519fbaae886e2773b37363970433d36cbb47d853/client/consensus/babe/src/lib.rs#L1172-L1176) when we need to verify the inherents of a child block.

it's not relevant here because when this is check the child is not activated yet and parent not deactivated yet?

Reply:
TBH, I'm pretty sure the stale mechanics weren't actually used and we don't 're-activate' the revert targets.

Reply:
That create inherent for parent blocks does not work either way. First I never witnessed this to actually happen in the logs and if it happened it would not work because the leaf would be already deactivated, but that is the case with or without this PR.

Reply:
Ok, this means I could have fixed something that was never actually used. I would still plea for merging until we know for sure. Better have correct unused code than incorrect unused code. :see_no_evil: 

https://github.com/paritytech/polkadot/issues/6097